Automattic/woocommerce-payments
# Store UPE appearance values in transient

Fixes #2161 

#### Changes proposed in this Pull Request

This PR adds the functionality to store the calculated UPE appearance values. The background is that to get UPE to visually match the store's theme, we need to do some "calculating" to build an object of CSS rules to pass in. While there is some performance benefit to not calculating that every time, the primary purpose of this PR is to make those calculated CSS rules available to other pages (pay for order, add payment method, update subscription payment method, etc) that use UPE also match the store's theme. The one caveat is that the store's checkout page must first be visited to cache these rules before they can be applied to the other pages mentioned.

This PR checks for the rules in a transient and passes what it finds into JS. In the JS it looks for the rules and if they are not set, calculates them and requests to save them on the server for future use.



#### Testing instructions

- Start with a store with UPE enabled and at least two themes installed. I used Storefront and Astra for my testing.
- With theme 1 selected, visit the Add Payment Method page with a logged in user. My Account > Payment Methods > Add Payment Method.
- Confirm that the UPE form loads, but the fields are not styled to match the theme (assuming checkout page has not been visited yet).
- Leave the add payment method screen without taking any action.
- Add an item to the cart and go to the checkout page. Confirm that the UPE fields are styled to match the theme on the checkout page.
- Without checking out, go back to the Add Payment Method page and confirm that the UPE fields are now styled to match the theme.
- Now as admin, switch to theme 2. This will clear the transient containing the appearance rules.
- Visit the Add Payment Method screen again and confirm that UPE is now back in it's original appearance (not matching either theme 1 or theme 2).
- Visit the checkout page and confirm that UPE is styled to match theme 2.
- Revisit the Add Payment Method screen and confirm UPE is now styled to match theme 2.

All of the above steps can also be taken using the Pay for Order page ([instructions](https://github.com/Automattic/woocommerce-payments/pull/2274)) or Update Subscription Pay Method page ([instructions](https://github.com/Automattic/woocommerce-payments/pull/2386)) to confirm the saved styles are being applied correctly.



**Unstyled Add Payment Method**
![Unstyled-UPE-AddPaymentMethod](https://user-images.githubusercontent.com/68524302/125514662-a25697de-69ce-436b-9c8a-70f621caea47.png)

**Visit Checkout page with styled UPE**
![Visit-Checkout-Styled-UPE](https://user-images.githubusercontent.com/68524302/125514671-f54cc1c7-b96b-4100-9378-0beec0083abc.png)

**Styled Add Payment Method**
![Styled-UPE-AddPaymentMethod](https://user-images.githubusercontent.com/68524302/125514684-a3c09647-9e6c-4b87-ab16-85f9b782a4e4.png)


-------------------
